core/server/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -0,0 +1,423 @@
+  protected abstract void initializeRequest(Protocol.ReadRequest request) throws Exception;
+
+  /**
+   * Returns the appropriate {@link DataBuffer} representing the data to send, depending on the
+   * configurable transfer type.
+   *
+   * @param channel the netty channel
+   * @param len The length, in bytes, of the data to read from the block
+   * @return a {@link DataBuffer} representing the data
+   * @throws IOException if an I/O error occurs when reading the data
+   */
+  protected abstract DataBuffer getDataBuffer(Channel channel, long offset, int len)
+      throws IOException;
+
+  /**
+   * The channel handler listener that runs after a packet write is flushed.
+   */
+  private final class WriteListener implements ChannelFutureListener {
+    private final long mPosToWriteUncommitted;
+
+    /**
+     * Creates an instance of the {@link WriteListener}.
+     *
+     * @param pos the pos
+     */
+    public WriteListener(long pos) {
+      mPosToWriteUncommitted = pos;
+    }
+
+    @Override
+    public void operationComplete(ChannelFuture future) {
+      if (!future.isSuccess()) {
+        LOG.error("Failed to send packet.", future.cause());
+        future.channel().pipeline().fireExceptionCaught(future.cause());
+        return;
+      }
+
+      mLock.lock();
+      try {
+        Preconditions.checkState(mPosToWriteUncommitted - mPosToWrite <= PACKET_SIZE,
+            "Some packet is not acked.");
+        incrementMetrics(mPosToWriteUncommitted - mPosToWrite);
+        mPosToWrite = mPosToWriteUncommitted;
+
+        if (shouldRestartPacketReader()) {
+          mPacketReaderExecutor.submit(new PacketReader(future.channel()));
+          mPacketReaderActive = true;
+        }
+        if (remainingToWrite() <= 0) {
+          replySuccess(future.channel());
+        }
+      } finally {
+        mLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * A runnable that reads from block worker and writes to the channel.
+   */
+  private class PacketReader implements Runnable {
+    Channel mChannel;
+
+    /**
+     * Creates an instance of the {@link PacketReader}.
+     *
+     * @param channel the channel
+     */
+    public PacketReader(Channel channel) {
+      mChannel = channel;
+    }
+
+    @Override
+    public void run() {
+      try {
+        runInternal();
+      } catch (Throwable e) {
+        LOG.error("Failed to run PacketReader.", e);
+        throw e;
+      }
+    }
+
+    /**
+     * The actual implementation of the runnable.
+     */
+    private void runInternal() {
+      while (true) {
+        final long start;
+        final int packet_size;
+        mLock.lock();
+        try {
+          start = mPosToQueue;
+          long remaining = remainingToQueue();
+          if (tooManyPendingPackets() || remaining <= 0) {
+            mPacketReaderActive = false;
+            break;
+          }
+
+          packet_size = (int) Math.min(remaining, PACKET_SIZE);
+          mPosToQueue += packet_size;
```

Comment:
This `mPosToQueue` is updated here, but the buffer is not read and sent until later (Exceptions are possible later too). Could an inconsistency arise?